waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,483 @@
+        base-body {:kind "DeleteOptions", :apiVersion "v1"}
+        term-json (-> base-body (assoc :gracePeriodSeconds 300) (as-json))
+        kill-json (-> base-body (assoc :gracePeriodSeconds 0) (as-json))
+        make-kill-response (fn [killed? message status]
+                             {:instance-id id, :killed? killed?,
+                              :message message, :service-id service-id, :status status})]
+    ; request termination of the instance
+    (api-request http-client pod-url :request-method :delete :body term-json)
+    ; scale down the replicaset
+    (scale-service-by api-server-url http-client service -1)
+    ; force-kill the instance (should still be terminating)
+    (api-request http-client pod-url :request-method :delete :body kill-json)
+    ; report back that the instance was killed
+    (scheduler/process-instance-killed! instance)
+    (make-kill-response true "pod was killed" 200)))
+
+(defn- service-spec
+  [service-id service-description service-id->password-fn]
+  (let [spec-file-path "./specs/k8s-default-pod.edn" ; TODO - allow user to provide this via the config file
+        {:strs [backend-proto cmd cpus #_disk grace-period-secs health-check-interval-secs
+                health-check-max-consecutive-failures mem ports min-instances
+                #_restart-backoff-factor run-as-user]} service-description
+        home-path (str "/home/" run-as-user)
+        common-env (scheduler/environment service-id service-description
+                                          service-id->password-fn home-path)
+        port0 8080
+        template-env (into [{:name "MESOS_DIR", :value home-path}
+                            {:name "MESOS_SANDBOX", :value home-path}]
+                           (concat
+                             (for [[k v] common-env]
+                               {:name k, :value v})
+                             (for [i (range ports)]
+                               {:name (str "PORT" i), :value (str (+ port0 i))})))
+        params {:app-name (service-id->app-name service-id)
+                :backend-protocol backend-proto
+                :backend-protocol-caps (string/upper-case backend-proto)
+                :cmd cmd
+                :cpus cpus
+                :env template-env
+                :grace-period-secs grace-period-secs
+                :health-check-interval-secs health-check-interval-secs
+                :health-check-max-consecutive-failures health-check-max-consecutive-failures
+                :health-check-url (sd/service-description->health-check-url service-description)
+                :home-path home-path
+                :memory  (str mem "Mi")
+                :min-instances min-instances
+                :port-count ports
+                :run-as-user run-as-user
+                :service-id service-id
+                :ssl? (= "https" backend-proto)}
+        delayed-values {:run-as-uid (->> run-as-user (shell/sh "id" "-u")
```

Comment:
Extract this into a function and maintain a cache?

Reply:
I'm just going to delete this for now. What I'd really like to add a way to register custom functions named in the config file.